packages/boot-contract-derive/src/lib.rs

```diff
@@ -52,12 +59,149 @@ pub fn contract(attrs: TokenStream, input: TokenStream) -> TokenStream {
             }
         }
 
-        impl<Chain: ::boot_core::CwEnv> ::boot_core::CwInterface for #name<Chain> {
+        impl<Chain: ::boot_core::CwEnv> ::boot_core::InstantiateableContract for #name<Chain> {
             type InstantiateMsg = #init;
+        }
+
+        impl<Chain: ::boot_core::CwEnv> ::boot_core::ExecuteableContract for #name<Chain> {
             type ExecuteMsg = #exec;
+        }
+
+        impl<Chain: ::boot_core::CwEnv> ::boot_core::QueryableContract for #name<Chain> {
             type QueryMsg = #query;
+        }
+
+        impl<Chain: ::boot_core::CwEnv> ::boot_core::MigrateableContract for #name<Chain> {
             type MigrateMsg = #migrate;
         }
     );
     struct_def.into()
 }
+
+#[proc_macro_attribute]
+pub fn boot_contract(_attrs: TokenStream, mut input: TokenStream) -> TokenStream {
+    let cloned = input.clone();
+    let mut item = parse_macro_input!(cloned as syn::Item);
+
+    let Item::Fn(boot_func) = &mut item else {
+        panic!("Only works on functions");
+    };
+
+    // Now we get the fourth function argument that should be the instantiate message
+    let signature = &mut boot_func.sig;
+    let func_ident = signature.ident.clone();
+    let func_type = get_func_type(signature);
+
+    let message_idx = match func_ident.to_string().as_ref() {
+        "instantiate" | "execute" => 3,
+        "query" | "migrate" => 2,
+        _ => panic!("Function name not supported for the macro"),
+    };
+
+    let message = match signature.inputs[message_idx].clone() {
+        FnArg::Typed(syn::PatType { ty, .. }) => *ty,
+        _ => panic!("Only typed arguments"),
+    };
+
+    let wasm_name = get_wasm_name();
+    let name = get_crate_to_struct();
+
+    let struct_def = quote!(
+            #[derive(
+                ::std::clone::Clone,
+            )]
+            pub struct #name<Chain: ::boot_core::CwEnv>(::boot_core::Contract<Chain>);
+
+            impl<Chain: ::boot_core::CwEnv> ::boot_core::ContractInstance<Chain> for #name<Chain> {
+                fn as_instance(&self) -> &::boot_core::Contract<Chain> {
+            &self.0
+        }
+            fn as_instance_mut(&mut self) -> &mut ::boot_core::Contract<Chain> {
+                &mut self.0
+            }
+        }
+
+        fn find_workspace_dir() -> ::std::path::PathBuf{
```

Comment:
Wondering about this. If the code that uses this package is published and imported by someone, what will happen? 

Essentially you either need to ship the wasms with the package (current boot-cw-plus approach) or you need the consumer of the package to have the wasms for the package in their own artifacts dir.

There are also two different uses, either the dev wants to publish his contracts individually and add the wasm to it individually, or the dev has a coherent set of contracts and wants to ship a package with a wrapper (ex. `CwPlus`) and all the wasms for the contracts. Which one do you want to support? Is there a way to support both?

Reply:
Here, we assume that the creator of the package ships the artifacts dir along with the crate. That is the approach you chose and I tried to respect that. I also think it is a sane approach for package creators, because the wasms are not heavy and you can check the wasm hash easily against an on-chain one if needed. 

To answer the second question, today this supports shipping a workspace (as this cw-plus version that is imported here) or a single contract repo. This allows for most simple implementations to create their interfaces easily and without too much boilerplate. 
For devs that want a wrapper with functions and handles they can develop it themselves (as for you boot-cw-plus version), because I see them as more experienced and able to ship such code. We could add in the future something to help wrap packages with nice function helpers

Reply:
The workspace approach works as long as the dependency is a git url to the repo. If we were to package one of the contracts and publish it to crates.io then your setup would break because the worspace artifacts wouldn't be included in that crate.

Reply:
And that's flaky behavior that people won't expect

Reply:
Another alternative could be adding a build.rs file that copies the workspace wasm to a dir in the crate to ensure it's included in the crate but that's a bit overkill